### PR TITLE
Feat: Neuron detail split

### PR DESCRIPTION
**Pull request by lmuntanerDfinity**
# Motivation

User can split the neuron in two.

# Changes

* Implement functionality of the Split Neuron button.
* Add splitNeuron service.
* Add splitNeuron api.
* New Modal to split the neuron by entering the amount.
* New neuron util neuronCanBeSplit.
* Upgrade nns-js.
* Upgrade dfinity agent, principal, and auth-client dependencies.

# Tests

* Test the new service
* Test the new api
* Test the new modal
* Test new neuron util

### CODE REVIEW

**Comment by lmuntanerDfinity** [comment]
New UI:

Screenshot 2022-04-08 at 08 27 46

**Comment by lmuntanerDfinity** [comment]
@peterpeterparker @mstrasinskis please review

File: frontend/svelte/src/lib/services/neurons.services.ts [hunk 1]
```diff
@@ -287,6 +300,53 @@ export const joinCommunityFund = async (neuronId: NeuronId): Promise<void> => {
...
+}): Promise<void> => {
+  try {
+    // Try/catch done in the component
+    await assertNeuronUserControlled(neuronId);
+
+    const identity: Identity = await getIdentity();
+
+    const stake = convertNumberToICP(amount);
+
+    if (stake === undefined) {
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/services/neurons.services.ts:318]: these two ifs to verify stake are duplicated in the component, worth a utils

**Review by peterpeterparker — COMMENTED** [review]
Morning, a couple of comments

File: frontend/svelte/src/lib/components/neuron-detail/actions/SplitNeuronButton.svelte [hunk 2]
```diff
@@ -1,5 +1,45 @@
...
+  const closeModal = () => (isOpen = false);
+
+  let isSplittable: boolean;
+  $: isSplittable = neuronCanBeSplit(neuron);
+  $i18n.neuron_detail.split_neuron_disabled_tooltip;
 </script>
 
-<button class="primary small">{$i18n.neuron_detail.split_neuron}</button>
+{#if isSplittable}
+  <button class="primary small" on:click={openModal}
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/neuron-detail/actions/SplitNeuronButton.svelte:24]: I don't find it interesting if clickable, but I can add it. And then yes, I can remove the duplicated code.

File: frontend/svelte/src/lib/constants/icp.constants.ts [hunk 3]
```diff
@@ -1,5 +1,7 @@
 export const E8S_PER_ICP = 100_000_000;
 export const TRANSACTION_FEE_E8S = 10_000;
+export const MIN_NEURON_STAKE_SPLITTABLE =
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/constants/icp.constants.ts:3]: sure.

File: frontend/svelte/src/lib/modals/neurons/SplitNeuronModal.svelte [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  let amount: number | undefined;
+
+  let stakeE8s: bigint;
+  $: stakeE8s = neuronStake(neuron);
+
+  let balance: ICP;
+  $: balance = ICP.fromE8s(stakeE8s);
+
+  let max: number = 0;
+  $: max = (Number(stakeE8s) - TRANSACTION_FEE_E8S) / E8S_PER_ICP;
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/SplitNeuronModal.svelte:28]: They could in an edge case. I agree with you, we better prepare for it.

File: frontend/svelte/src/lib/modals/neurons/SplitNeuronModal.svelte [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+  let validForm: boolean;
+  $: validForm = amount !== undefined && amount > 0 && amount <= max;
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/SplitNeuronModal.svelte:31]: I didn't think it was needed since it's a simple logic easy to read. But I can create the util.

File: frontend/svelte/src/lib/modals/neurons/SplitNeuronModal.svelte [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+  const onMax = () => (amount = max);
+
+  const dispatcher = createEventDispatcher();
+  const split = async () => {
+    // TS is not smart enought to understand that `validForm` also covers `amount === undefined`
+    if (!validForm || amount === undefined) {
+      return;
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/SplitNeuronModal.svelte:39]: Yes, good point. It's an edge case, but rather be safe.

File: frontend/svelte/src/lib/services/neurons.services.ts [hunk 7]
```diff
@@ -78,6 +79,19 @@ export const assertNeuronControllable = async (
   }
 };
 
+const convertNumberToICP = (amount: number): ICP | undefined => {
+  const stake = ICP.fromString(String(amount));
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/neurons.services.ts:83]: sure!

File: frontend/svelte/src/lib/services/neurons.services.ts [hunk 8]
```diff
@@ -287,6 +300,53 @@ export const joinCommunityFund = async (neuronId: NeuronId): Promise<void> => {
...
 
+export const splitNeuron = async ({
+  neuronId,
+  amount,
+}: {
+  neuronId: NeuronId;
+  amount: number;
+}): Promise<void> => {
+  try {
+    // Try/catch done in the component
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/neurons.services.ts:311]: yes, that was my initial idea, and then I didn't like it. Thanks for catching it!

File: frontend/svelte/src/lib/services/neurons.services.ts (hunk 1 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/neurons.services.ts:318]: Do you mean two utils or one utils with two ifs?

File: frontend/svelte/src/lib/components/neuron-detail/actions/SplitNeuronButton.svelte [hunk 9]
```diff
@@ -1,5 +1,45 @@
...
+  export let neuron: NeuronInfo;
+
+  let isOpen: boolean = false;
+
+  const openModal = () => (isOpen = true);
+  const closeModal = () => (isOpen = false);
+
+  let isSplittable: boolean;
+  $: isSplittable = neuronCanBeSplit(neuron);
+  $i18n.neuron_detail.split_neuron_disabled_tooltip;
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/neuron-detail/actions/SplitNeuronButton.svelte:20]: to be removed, forgot to remove it.

File: frontend/svelte/src/lib/services/neurons.services.ts [hunk 10]
```diff
@@ -287,6 +300,53 @@ export const joinCommunityFund = async (neuronId: NeuronId): Promise<void> => {
...
+        labelKey: "error.amount_not_enough",
+      });
+      return;
+    }
+    await splitNeuronApi({ neuronId, identity, amount: stake });
+    toastsStore.show({
+      labelKey: "neuron_detail.split_neuron_success",
+      level: "info",
+    });
+    // TODO: Should we try/catch this independently? Or have it inside?
```
Comment (lmuntanerDfinity) [inline frontend/svelte/src/lib/services/neurons.services.ts:336]: I will work in a separate PR on how to handle errors in the neuron details page functionality.

Now I have enough use cases to find the common pattern and create a good abstraction.

File: frontend/svelte/src/lib/services/neurons.services.ts (hunk 1 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/neurons.services.ts:318]: Since this needs to return (for now), not just throw (which would cause a second toast error. I will manage this code repetition separately.

I mention below that managing errors needs to be improved and now that I have enough use cases we should be able to find a good abstraction.

**Comment by lmuntanerDfinity** [comment]
@peterpeterparker ready for another review!

File: frontend/svelte/src/lib/services/neurons.services.ts (hunk 1 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/services/neurons.services.ts:305]: I mean the exact same piece of code `if (stake === undefined) {....` and `if (stake.toE8s() < E8S_PER_ICP) {...` find place twice in the code. Once here in the service and once in the component.

Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/neurons.services.ts:305]: I need to return inside the `if`, do you mean to create an util that returns a boolean instead of having `stake === undefined`, and another with `stake.toE8s() < E8S_PER_ICP`?

For the one below makes sense. `isEnoughToStakeNeuron(stake)`.

I was thinking on implementing custom errors and manage all this in the `catch` and then reuse the locig to show one error or another with the toast depending on the Custom Error. WDYT?

File: frontend/svelte/src/lib/components/accounts/NewTransactionAmount.svelte [hunk 11]
```diff
@@ -24,20 +27,26 @@
...
+  let balance: ICP;
+  $: ({ balance } = $store.selectedAccount ?? {
+    balance: ICP.fromE8s(BigInt(0)),
+  });
 
   const onMax = () => (amount = max);
   const onSubmit = () => {
-    if (!validForm) {
+    // TS not smart enough to know that validForm also covers `amount === undefiend`
+    if (!validForm || amount === undefined) {
```
Comment (mstrasinskis) [inline frontend/svelte/src/lib/components/accounts/NewTransactionAmount.svelte:40]: `isValidInputAmount` already checks for undefined.

File: frontend/svelte/src/lib/utils/neuron.utils.ts [hunk 12]
```diff
@@ -168,3 +169,27 @@ export const neuronStake = (neuron: NeuronInfo): bigint =>
   neuron.fullNeuron?.cachedNeuronStake !== undefined
     ? neuron.fullNeuron?.cachedNeuronStake - neuron.fullNeuron?.neuronFees
     : BigInt(0);
+
+export const neuronCanBeSplit = (neuron: NeuronInfo): boolean =>
+  neuronStake(neuron) >= BigInt(MIN_NEURON_STAKE_SPLITTABLE);
```
Comment (mstrasinskis) [inline frontend/svelte/src/lib/utils/neuron.utils.ts:174]: Currently in the Flutter the minimum is `1.0001` ICP. Could you please recheck it in the code?
(`frontend/dart/lib/ui/transaction/wallet/split_neuron_amount_page.dart`)


image

**Review by mstrasinskis — COMMENTED** [review]
A couple of comments to the draft.

File: frontend/svelte/src/lib/components/accounts/NewTransactionAmount.svelte (hunk 11 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/accounts/NewTransactionAmount.svelte:40]: I know, but Typescript is not smart enough. Hence the comment above.

File: frontend/svelte/src/lib/utils/neuron.utils.ts (hunk 12 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/utils/neuron.utils.ts:174]: I hadn't seen this validation. But in the button is the following:

```
.takeIf((e) =>
              neuron.isCurrentUserController &&
              neuron.balance.asE8s() >=
                  BigInt.from((2 * E8S_PER_ICP) + TRANSACTION_FEE_E8S)),
```

Hence that I used it for my validation.

The validation you saw is for the input, I will add it there. It's the same validation as staking a neuron.

Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/utils/neuron.utils.ts:174]: This is for the new neuron stake. I changed the helper to be able to take into account the transaction fee.

**Comment by lmuntanerDfinity** [comment]
@peterpeterparker @mstrasinskis ready for another review

File: frontend/svelte/src/lib/services/neurons.services.ts (hunk 1 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/services/neurons.services.ts:306]: actually not in the component, in the services:

`.stakeAndLoadNeuron()`:

```
if (stake === undefined) {
      toastsStore.error({
        labelKey: "error.amount_not_valid",
      });
      return;
    }

    if (!isEnoughToStakeNeuron({ stake })) {
      toastsStore.error({
        labelKey: "error.amount_not_enough",
      });
      return;
    }
    ```

`.splitNeuron()`:

```
if (stake === undefined) {
      toastsStore.error({
        labelKey: "error.amount_not_valid",
      });
      return;
    }

    if (!isEnoughToStakeNeuron({ stake, withTransactionFee: true })) {
      toastsStore.error({
        labelKey: "error.amount_not_enough",
      });
      return;
    }
    ```
    
    except `withTransactionFee` it's really similar

Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/neurons.services.ts:306]: I know, I will change both of these to raise errors and handle them in the catch. I think it will make it easier than having all this here.

I will refactor all the error management in the neuron functions in another PR, unifying them all. This also.

Comment (peterpeterparker) [reply frontend/svelte/src/lib/services/neurons.services.ts:306]: Sounds good 👍

**Comment by lmuntanerDfinity** [comment]
@peterpeterparker @mstrasinskis please review

File: frontend/svelte/package-lock.json [hunk 13]
```diff
@@ -8,9 +8,10 @@
       "name": "svelte-app",
       "version": "1.0.0",
       "dependencies": {
-        "@dfinity/auth-client": "^0.10.3",
-        "@dfinity/identity": "^0.10.3",
-        "@dfinity/nns": "^0.3.0"
+        "@dfinity/agent": "^0.11.0",
+        "@dfinity/auth-client": "^0.11.0",
+        "@dfinity/nns": "^0.3.0-nightly-2022-04-12",
+        "@dfinity/principal": "^0.11.0"
```
Comment (peterpeterparker) [inline frontend/svelte/package-lock.json:14]: why do we need `@dfinity/principal` as an explicit dependency, is that related to what you add in your PR?

File: frontend/svelte/package-lock.json [hunk 14]
```diff
@@ -680,71 +681,108 @@
...
+        "@dfinity/candid": "^0.11.0",
+        "@dfinity/principal": "^0.11.0",
+        "crc": "^4.1.1",
+        "crc-32": "^1.2.2",
+        "google-protobuf": "^3.20.0",
         "js-sha256": "^0.9.0",
         "randombytes": "^2.1.0"
       }
     },
+    "node_modules/@dfinity/nns/node_modules/buffer": {
```
Comment (peterpeterparker) [inline frontend/svelte/package-lock.json:747]: this new dependency and `"node_modules/@dfinity/nns/node_modules/crc"`  are related to the `@dfinity/principal`?

File: frontend/svelte/package-lock.json (hunk 13 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/package-lock.json:14]: Maybe it's not needed, let me check.

Comment (lmuntanerDfinity) [reply frontend/svelte/package-lock.json:14]: I removed it from `package.json`, but it didn't get removed from here.

Anyway, I find it also weird that we don't have it as a dependency but we import directly `import { Principal } from "@dfinity/principal";`

Comment (lmuntanerDfinity) [reply frontend/svelte/package-lock.json:14]: Maybe, but this PR also needed the upgrade on nns-js, which needed the upgrade in the rest. The principal can be done in a separate PR as you say

File: frontend/svelte/package-lock.json (hunk 14 above)
Comment (peterpeterparker) [reply frontend/svelte/package-lock.json:747]: Double checked it's correct, follow libs that have been bumped in nns-js
